dali/kernels/slice/slice_flip_normalize_permute_kernel_test.h

```diff
@@ -0,0 +1,439 @@
+template <typename TestArgs>
+class SliceFlipNormalizePermuteTest : public ::testing::Test {
+ public:
+  using InputType = typename TestArgs::InputType;
+  using OutputType = typename TestArgs::OutputType;
+  static constexpr size_t Dims = TestArgs::Dims;
+  static constexpr size_t NumSamples = TestArgs::NumSamples;
+  static constexpr size_t DimSize = TestArgs::DimSize;
+  static constexpr size_t DimSize0 = TestArgs::DimSize0;
+  static constexpr size_t DimSize1 = TestArgs::DimSize1;
+  using ArgsGenerator = typename TestArgs::ArgsGenerator;
+  using KernelArgs = SliceFlipNormalizePermuteArgs<OutputType, Dims>;
+
+  void PrepareData(TestTensorList<InputType, Dims>& test_data) {
+    std::vector<int> sample_dims(Dims, DimSize);
+    sample_dims[0] = DimSize0;
+    sample_dims[1] = DimSize1;
+    TensorListShape<Dims> shape = uniform_list_shape<Dims>(NumSamples, sample_dims);
+    test_data.reshape(shape);
+
+    InputType num = 0;
+    auto seq_gen = [&num]() { return num++; };
+    Fill(test_data.cpu(), seq_gen);
+  }
+
+  void PrepareExpectedOutput(TestTensorList<InputType, Dims>& input_data,
+                             std::vector<KernelArgs>& args,
+                             TestTensorList<OutputType, Dims>& output_data) {
+    auto in = input_data.cpu();
+    std::vector<TensorShape<Dims>> output_shapes;
+    std::vector<TensorShape<Dims>> slice_shapes;
+    for (int i = 0; i < in.size(); i++) {
+      auto shape = args[i].shape;
+      slice_shapes.push_back(shape);
+      auto padded_shape = args[i].should_pad ?
+        args[i].padded_shape : shape;
+      auto out_shape = padded_shape;
+      for (size_t d = 0; d < Dims; d++) {
+        auto perm_d = args[i].should_permute ? args[i].permuted_dims[d] : d;
+        out_shape[d] = padded_shape[perm_d];
+      }
+      TensorShape<Dims> out_sample_shape(out_shape);
+      output_shapes.push_back(out_sample_shape);
+    }
+    output_data.reshape(output_shapes);
+    auto out = output_data.cpu();
+
+    for (int i = 0; i < in.size(); i++) {
+      const auto& anchor = args[i].anchor;
+      const auto& flip = args[i].flip;
+      const auto *in_tensor = in.tensor_data(i);
+      auto *out_tensor = out.tensor_data(i);
+      const auto &permuted_dims = args[i].permuted_dims;
+      auto slice_shape = slice_shapes[i];
+      auto should_flip = args[i].should_flip;
+      auto should_permute = args[i].should_permute;
+      auto should_normalize = args[i].should_normalize;
+      auto should_pad = args[i].should_pad;
+
+      const auto in_shape = in.tensor_shape(i);
+      std::array<int64_t, Dims> in_strides = GetStrides<Dims>(in_shape);
+
+      const auto out_shape = out.tensor_shape(i);
+      std::array<int64_t, Dims> out_strides = GetStrides<Dims>(out_shape);
+
+      // normalization
+      auto mean = args[i].mean;
+      auto inv_stddev = args[i].inv_stddev;
+      // reverse channels if last dim flip was requested
+      if (should_normalize && should_flip && flip[Dims - 1]) {
```

Comment:
Shouldn't you reverse mean and std_dev only when `flip[normalize_dim]` is set?

Reply:
done